fluent-bit/fluent-bit-go-loki/out_loki.go

```diff
@@ -0,0 +1,183 @@
+	Exit(code int)
+}
+
+type fluentPlugin struct{}
+
+func (p *fluentPlugin) PluginConfigKey(ctx unsafe.Pointer, key string) string {
+	return output.FLBPluginConfigKey(ctx, key)
+}
+
+func (p *fluentPlugin) Unregister(ctx unsafe.Pointer) {
+	output.FLBPluginUnregister(ctx)
+}
+
+func (p *fluentPlugin) GetRecord(dec *output.FLBDecoder) (int, interface{}, map[interface{}]interface{}) {
+	return output.GetRecord(dec)
+}
+
+func (p *fluentPlugin) NewDecoder(data unsafe.Pointer, length int) *output.FLBDecoder {
+	return output.NewDecoder(data, length)
+}
+
+func (p *fluentPlugin) Exit(code int) {
+	os.Exit(code)
+}
+
+func (p *fluentPlugin) HandleLine(ls model.LabelSet, timestamp time.Time, line string) error {
+	return loki.Handle(ls, timestamp, line)
+}
+
+//export FLBPluginRegister
+func FLBPluginRegister(ctx unsafe.Pointer) int {
+	return output.FLBPluginRegister(ctx, "loki", "Ship fleunt-bit logs to Grafana Loki")
+}
+
+//export FLBPluginInit
+// (fluentbit will call this)
+// ctx (context) pointer to fluentbit context (state/ c code)
+func FLBPluginInit(ctx unsafe.Pointer) int {
+	// Example to retrieve an optional configuration parameter
+	url := plugin.PluginConfigKey(ctx, "URL")
+	batchWait := plugin.PluginConfigKey(ctx, "BatchWait")
+	batchSize := plugin.PluginConfigKey(ctx, "BatchSize")
+	labels := plugin.PluginConfigKey(ctx, "Labels")
+
+	config, err := getLokiConfig(url, batchWait, batchSize, labels)
+	if err != nil {
+		plugin.Unregister(ctx)
+		plugin.Exit(1)
+		return output.FLB_ERROR
+	}
+	fmt.Printf("[flb-go] plugin version is %s\n", Version)
+	fmt.Printf("[flb-go] plugin URL parameter = '%s'\n", url)
+	fmt.Printf("[flb-go] plugin BatchWait parameter = '%s'\n", batchSize)
+	fmt.Printf("[flb-go] plugin BatchSize parameter = '%s'\n", batchWait)
+	fmt.Printf("[flb-go] plugin Labels parameter = '%s'\n", labels)
+
+	cfg := client.Config{}
+	// Init everything with default values.
+	flagext.RegisterFlags(&cfg)
+
+	// Override some of those defaults
+	cfg.URL = config.url
+	cfg.BatchWait = config.batchWait
+	cfg.BatchSize = config.batchSize
+
+	log := logrus.New()
+
+	loki, err = client.New(cfg, kit.NewLogrusLogger(log))
+	if err != nil {
+		log.Fatalf("client.New: %s\n", err)
+		plugin.Unregister(ctx)
+		plugin.Exit(1)
+		return output.FLB_ERROR
+	}
+	ls = config.labelSet
+
+	return output.FLB_OK
+}
+
+//export FLBPluginFlush
+func FLBPluginFlush(data unsafe.Pointer, length C.int, tag *C.char) int {
+	var ret int
+	var ts interface{}
+	var record map[interface{}]interface{}
+
+	dec := plugin.NewDecoder(data, int(length))
+
+	for {
+		ret, ts, record = plugin.GetRecord(dec)
+		if ret != 0 {
+			break
+		}
+
+		// Get timestamp
+		var timestamp time.Time
+		switch t := ts.(type) {
+		case output.FLBTime:
+			timestamp = ts.(output.FLBTime).Time
+		case uint64:
+			timestamp = time.Unix(int64(t), 0)
+		default:
+			fmt.Print("timestamp isn't known format. Use current time.")
+			timestamp = time.Now()
+		}
+
+		line, err := createJSON(record)
```

Comment:
There is something wrong here Loki is supposed to received only the the actual text as line. It seems that you are also embedding metadata in it using json.

I think you should translate records into labels (except for the line part) and merge it with the global labels values from the configuration.

Most people will use that fluentbit with the kubernetes plugin so discovery will be done that way and the record will contain information like pod_name, container_id ..... and those should not be part of the line sent to Loki.

Reply:
see https://fluentbit.io/documentation/0.12/filter/kubernetes.html

Reply:
check our fluentd configuration https://github.com/grafana/loki/tree/master/fluentd/fluent-plugin-grafana-loki#output-format

That should give you a sense of what is required.

Reply:
fluent-bit has loosely coupled plugin ecosystem and they loves to single functional plugin.
Users should consider to filter record with another plugin (such as filter grep) in their fluent-bit data pipeline:
https://docs.fluentbit.io/manual/v/1.2/filter/grep

IMO, I don't want to add record manipulation functionality in this plugin.
But it is OK to add.

Reply:
I think we should support the same config as fluentd plugin except `drop_single_key`. You're very close you just need those 3:

remove_keys: (default: nil) comma separated list of needless record keys to remove. All other keys will be placed into the log line

label_keys: (default: "job,instance") comma separated list of keys to use as stream labels. All other keys will be placed into the log line

line_format: format to use when flattening the record to a log line. Valid values are "json" or "key_value". If set to "json" the log line sent to Loki will be the fluentd record (excluding any keys extracted out as labels) dumped as json. If set to "key_value", the log line will be each item in the record concatenated together (separated by a single space) in the format <key>=<value>.